test/e2e/storage/subpath.go

```diff
@@ -168,203 +137,373 @@ var _ = utils.SIGDescribe("Subpath", func() {
-				setWriteCommand(subPathDir, &pod.Spec.Containers[1])
+		return subPathTestInput{
+			f:                 f,
+			subPathDir:        subPathDir,
+			filePathInSubpath: filepath.Join(volumePath, fileName),
+			filePathInVolume:  filepath.Join(subPathDir, fileName),
+			volType:           r.volType,
+			pod:               r.pod,
+			volSource:         r.volSource,
+			roVol:             r.roVolSource,
+		}
+	}
 
-				// Read it from inside the subPath from container 0
-				pod.Spec.Containers[0].VolumeMounts[0].ReadOnly = true
-				testReadFile(f, volumePath, pod, 0)
-			})
+	framework.Failf("Fail to convert testResource to subPathTestResource")
+	return nil
+}
 
-			It("should support existing directories when readOnly specified in the volumeSource", func() {
-				roVol := vol.getReadOnlyVolumeSpec()
-				if roVol == nil {
-					framework.Skipf("Volume type %v doesn't support readOnly source", curVolType)
-				}
+func (s *subPathTestSuite) getTestFunc(testPattern testPattern, driver testDriver) func(*testInput) {
+	return testSubPathWithTestInput
+}
 
-				// Initialize content in the volume while it's writable
-				initVolumeContent(f, pod, filePathInVolume, filePathInSubpath)
+func testSubPathWithTestInput(testInput *testInput) {
+	Context("-", func() {
+		var (
+			t  subPathTestInput
+			ok bool
+		)
 
-				// Set volume source to read only
-				pod.Spec.Volumes[0].VolumeSource = *roVol
+		BeforeEach(func() {
+			testInputVal := *testInput
+			if t, ok = testInputVal.(subPathTestInput); !ok {
+				framework.Failf("Fail to convert testInput to subPathTestInput")
+			}
+		})
 
-				// Read it from inside the subPath from container 0
-				testReadFile(f, filePathInSubpath, pod, 0)
-			})
+		testSubPath(&t)
+	})
+}
 
-			It("should fail for new directories when readOnly specified in the volumeSource", func() {
-				roVol := vol.getReadOnlyVolumeSpec()
-				if roVol == nil {
-					framework.Skipf("Volume type %v doesn't support readOnly source", curVolType)
+type subPathTestResource struct {
+	driver      testDriver
+	volType     string
+	volSource   *v1.VolumeSource
+	pvc         *v1.PersistentVolumeClaim
+	pv          *v1.PersistentVolume
+	roVolSource *v1.VolumeSource
+	sc          *storagev1.StorageClass
+	pod         *v1.Pod
+}
+
+func (s *subPathTestResource) setup(driver testDriver, testPattern testPattern) {
+	s.driver = driver
+	driverInfo := s.driver.getDriverInfo()
+	f := driverInfo.f
+	cs := f.ClientSet
+	fsType := testPattern.fsType
+
+	switch testPattern.testVolType {
+	case inlineVolume:
+		framework.Logf("Creating resouce for inline volume")
+		if inlineVolumeTestDriver, ok := driver.(inlineVolumeTestDriver); ok {
+			s.volType = driverInfo.name
+			s.volSource = inlineVolumeTestDriver.getVolumeSource(false, fsType)
+			s.roVolSource = inlineVolumeTestDriver.getVolumeSource(true, fsType)
+		}
+	case staticPV:
+		framework.Logf("Creating resouce for static PV")
+		if staticPVTestDriver, ok := driver.(staticPVTestDriver); ok {
```

Comment:
How come the static and dynamic setup cannot reuse what's already done in `genericVolumeTestResource.setup()`?

Reply:
Subpath test can share generic resource if below two are resolved:
 
1. Subpath test requires ro volume source. For inline volume, readOnly flag is embeded inside specific VolumeSource's struct, so its difficult to override it.
  https://github.com/kubernetes/kubernetes/blob/9e239b1986509458d4dd4896724ea1c703ec0c79/test/e2e/storage/subpath.go#L246

2. Subpath test has pod deletion logic outside It(), so pod is defined in resource to delete there. 
  https://github.com/kubernetes/kubernetes/blob/9e239b1986509458d4dd4896724ea1c703ec0c79/test/e2e/storage/subpath.go#L307

If we skip inline volume test for subpath, 1. will be resolved, so only 2. needs to be considered.

Reply:
Hm is it possible to call `genericVolumeTestResource.setup()` manually, and then do the additional calls needed for 1 + 2 in the `subpath.setup()`?

Reply:
Oh, I understand your intention, now. I will reflect it. Thank you for your suggestion!